components/eventing-controller/pkg/backend/jetstreamv2/jetstream.go

```diff
@@ -387,81 +389,112 @@ func (js *JetStream) deleteConsumerFromJetStream(name string) error {
+// syncNATSConsumers makes sure there are consumers and subscriptions created on the NATS Backend.
+func (js *JetStream) syncNATSConsumers(subscription *eventingv1alpha2.Subscription, asyncCallback func(m *nats.Msg)) error {
 	for _, subject := range subscription.Status.Types {
 		jsSubject := js.getJetStreamSubject(subscription.Spec.Source, subject.CleanType, subscription.Spec.TypeMatching)
 		jsSubKey := NewSubscriptionSubjectIdentifier(subscription, jsSubject)
-		log := log.With("subject", subject)
+		log := backendutilsv2.LoggerWithSubscriptionAndEventType(js.namedLogger(), subscription, subject)
 
-		if existingNatsSub, ok := js.subscriptions[jsSubKey]; !ok {
-			continue
-		} else if existingNatsSub.IsValid() {
-			log.Debugw("Skipping creation subscription on JetStream because it already exists")
-			continue
-		}
-		log.Debugw("Recreating subscription on JetStream because it was invalid")
-		// bind the existing consumer to a new subscription on JetStream
-		jsSubscription, err := js.jsCtx.Subscribe(
-			jsSubject,
-			asyncCallback,
-			nats.Bind(js.Config.JSStreamName, jsSubKey.ConsumerName()),
-		)
+		consumerInfo, err := js.jsCtx.ConsumerInfo(js.Config.JSStreamName, jsSubKey.ConsumerName())
 		if err != nil {
-			if err != nats.ErrConsumerNotFound {
-				log.Errorw("Failed to bind subscription to an existing consumer", "error", err)
+			// create the consumer in case it doesn't exist
+			if errors.Is(err, nats.ErrConsumerNotFound) {
+				consumerInfo, err = js.jsCtx.AddConsumer(
+					js.Config.JSStreamName,
+					js.getConsumerConfig(subscription, jsSubKey, jsSubject),
+				)
+				if err != nil {
+					return fmt.Errorf("%w: %v", ErrConsumerAdd, err)
+				}
+				log.Debug("Created consumer on JetStream")
+			} else {
+				return fmt.Errorf("%w: %v", ErrConsumerInfo, err)
 			}
-			delete(js.subscriptions, jsSubKey)
-		} else {
-			// save recreated JetStream subscription in storage
-			js.subscriptions[jsSubKey] = &Subscription{Subscription: jsSubscription}
-			log.Debugw("Recreated subscription on JetStream")
 		}
+
+		natsSubscription, subExists := js.subscriptions[jsSubKey]
+
+		// try to create a NATS Subscription if it doesn't exist
+		if !subExists && !consumerInfo.PushBound {
+			if err := js.createNATSSubscription(subscription, subject, jsSubject, jsSubKey, asyncCallback, log); err != nil {
+				return err
+			}
+		}
+
+		// try to bind invalid NATS Subscriptions
+		if subExists && !natsSubscription.IsValid() {
+			if err := js.bindConsumersForInvalidNATSSubscriptions(jsSubject, jsSubKey, asyncCallback, log); err != nil {
+				return err
+			}
+		}
+
+		if consumerInfo != nil {
+			if err := js.checkSubscriptionConfig(subscription, *consumerInfo, log); err != nil {
+				return err
+			}
+		}
+	}
+	if err := js.checkNATSSubscriptionsCount(subscription); err != nil {
```

Comment:
Why do we need to check this ? Isn't is the responsibility of this method to create them ? 🤔

Reply:
that's coming from https://github.com/kyma-project/kyma/pull/15447

Reply:
I see it was refactored in the new version. Thanks for this 👍